File: agent/python/mango_agent/thought_sanitize.py
from __future__ import annotations

import re

# Qwen/Gemma and similar models emit … wrappers.
_THINK_TAG = r"(?:redacted_thinking|think(?:ing)?)"
_THINK_BLOCK_RE = re.compile(
    rf"<\s*{_THINK_TAG}\b[^>]*>([\s\S]*?)<\s*/\s*{_THINK_TAG}\s*>",
    re.IGNORECASE,
)
_THINK_OPEN_OR_CLOSE_RE = re.compile(rf"<\s*/?\s*{_THINK_TAG}\b[^>]*>", re.IGNORECASE)
_THINK_PARTIAL_END_RE = re.compile(rf"<\s*/?\s*{_THINK_TAG}\b[^>]*$", re.IGNORECASE)
_CHANNEL_RE = re.compile(r"<\|?channel\|?>", re.IGNORECASE)
_FENCE_RE = re.compile(r"```[\w+-]*\n[\s\S]*?```", re.MULTILINE)


def strip_thought_markup(text: str) -> str:
    if not text:
        return ""
    cut = re.search(r"<tool_call\b", text, flags=re.IGNORECASE)
    head = text[: cut.start()] if cut else text
    dump = re.search(r"[A-Za-z_][\w.]*\([^)]*\)\s*\|", head)
    until_dump = head[: dump.start()] if dump else head
    cleaned = _CHANNEL_RE.sub("", until_dump)
    cleaned = _THINK_BLOCK_RE.sub(r"\1", cleaned)
    cleaned = _THINK_OPEN_OR_CLOSE_RE.sub("", cleaned)
    cleaned = _THINK_PARTIAL_END_RE.sub("", cleaned)
    cleaned = re.sub(r"<[^>]*$", "", cleaned)
    cleaned = _FENCE_RE.sub("", cleaned)
    cleaned = re.sub(r"<tool(?:_call\b[\s\S]*)?$", "", cleaned, flags=re.IGNORECASE)
    cleaned = re.sub(r"[ \t]+\n", "\n", cleaned)
    cleaned = re.sub(r"\n{3,}", "\n\n", cleaned)
    return cleaned.strip()


def is_empty_thought(text: str) -> bool:
    return not strip_thought_markup(text).strip()

File: apps/electron/src/renderer/src/lib/thoughtSanitize.ts
/** Strip model think wrappers and tool junk from visible thought text. */

const THINK_TAG = "(?:redacted_thinking|think(?:ing)?)";
const THINK_BLOCK_RE = new RegExp(
  `<\\s*${THINK_TAG}\\b[^>]*>([\\s\\S]*?)<\\s*\\/\\s*${THINK_TAG}\\s*>`,
  "gi",
);
const THINK_OPEN_OR_CLOSE_RE = new RegExp(`<\\s*\\/?\\s*${THINK_TAG}\\b[^>]*>`, "gi");
const THINK_PARTIAL_END_RE = new RegExp(`<\\s*\\/?\\s*${THINK_TAG}\\b[^>]*$`, "i");
const CHANNEL_RE = /<\|?channel\|?>/gi;

export function stripThoughtMarkup(text: string): string {
  const cut = text.search(/<tool_call\b/i);
  const head = cut >= 0 ? text.slice(0, cut) : text;
  const dump = head.search(/[A-Za-z_][\w.]*\([^)]*\)\s*\|/);
  const untilDump = dump >= 0 ? head.slice(0, dump) : head;
  let cleaned = untilDump
    .replace(CHANNEL_RE, "")
    .replace(THINK_BLOCK_RE, "$1")
    .replace(THINK_OPEN_OR_CLOSE_RE, "")
    .replace(THINK_PARTIAL_END_RE, "")
    .replace(/<[^>]*$/i, "")
    .replace(/```[\w+-]*\n[\s\S]*?```/g, "")
    .replace(/<tool(?:_call\b[\s\S]*)?$/i, "")
    .replace(/[ \t]+\n/g, "\n")
    .replace(/\n{3,}/g, "\n\n")
    .trim();
  return cleaned;
}

export function isEmptyThought(text: string): boolean {
  return !stripThoughtMarkup(text).trim();
}
